guardian/cdk
# ci: Perform project generation in CI

This is related to https://github.com/guardian/cdk/pull/1031.

## What does this change?


In this change, we run the new project generator CLI command (`npx @guardian/cdk new`) during CI. This should help us spot any regressions early as CI would fail.

Examples of regressions include:
  - Renaming of properties to a `GuStack`
  - Addition of mandatory properties to a `GuStack` (such as in https://github.com/guardian/cdk/pull/1134 which adds a new `stage` prop)
  - Upgrading to AWS CDK v2 (#1150)

The process is:
  1. CI checkouts the branch
  2. [`npm link`](https://docs.npmjs.com/cli/v8/commands/npm-link) the branch being built
  3. Create and move into a new random folder in `/tmp`
  4. Install `@guardian/cdk` via the `npm link` (i.e. from the current branch)
  5. Use the GuCDK CLI to generate a new project
    - The CLI has been updated to omit `@guardian/cdk` from the generated `package.json`, to ensure we use version from the current branch, and not a published version from `npm`
  6. Run the generated `./script/ci` bash script

The generated `ci` bash script [includes](https://github.com/guardian/cdk/blob/main/src/bin/commands/template/script/ci) a load of `yarn` commands. This has potential for confusion as this repo [explicitly chooses](https://github.com/guardian/cdk/blob/main/docs/architecture-decision-records/003-package-manager.md) to use npm.

## How to test



We want to test the generated project installs `@guardian/cdk` from the current branch and not NPM.

A commit has been added to this branch to `throw` within a `GuStack`. If we can see this exception during CI, then we know we've installed from the branch.

We do ([build log](https://github.com/guardian/cdk/runs/5508025429?check_suite_focus=true)):
<img width="602" alt="image" src="https://user-images.githubusercontent.com/836140/157829687-ed6d3177-6707-4991-90f1-d5bd1dfd99ef.png">

## How can we measure success?


Spotting regressions early, and before they are shipped.

## Have we considered potential risks?


Build times increase that little bit more 😢 .

## Checklist


[^1]: Consider whether this is something that will mean changes to projects that have already been migrated, or to the CDK CLI tool. If changes are required, consider adding a checklist here and/or linking to related PRs.
[^2]: If you are adding a new construct or pattern, has new documentation been added? If you are amending defaults or changing behaviour, are the existing docs still valid?
